waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/mapping/service/impl/PrefixBasedDatabaseMappingService.java

```diff
@@ -252,62 +261,143 @@ private boolean isWhitelisted(String databasePrefix, String database) {
+        for (Future<List<TableMeta>> future : futures) {
+          try {
+            DatabaseMapping mapping = iterator.next();
+            long timeout = mapping.getTimeout();
+            List<TableMeta> mappedTableMeta = future.get(timeout, TimeUnit.MILLISECONDS);
+            combined.addAll(mappedTableMeta);
+          } catch (InterruptedException | ExecutionException | TimeoutException | NullPointerException e) {
+            LOG.warn(errorMessage, e);
           }
         }
+        shutdownExecutorService(executorService);
         return combined;
       }
 
       @Override
       public List<String> getAllDatabases(String databasePattern) {
         List<String> combined = new ArrayList<>();
-        for (Entry<DatabaseMapping, String> mappingWithPattern : databaseMappingsByDbPattern(databasePattern)
-            .entrySet()) {
-          try {
-            DatabaseMapping mapping = mappingWithPattern.getKey();
-            String pattern = mappingWithPattern.getValue();
-            List<String> databases = mapping.getClient().get_databases(pattern);
-            for (String database : databases) {
-              if (isWhitelisted(mapping.getDatabasePrefix(), database)) {
-                combined.add(mapping.transformOutboundDatabaseName(database));
-              }
+        Map<DatabaseMapping, String> databaseMappingsForPattern = databaseMappingsByDbPattern(databasePattern);
+
+        ExecutorService executorService = Executors.newFixedThreadPool(databaseMappingsForPattern.size());
+        List<Future<List<String>>> futures = new ArrayList<>();
+        String errorMessage = "Can't fetch databases by pattern: ";
+
+        for (Entry<DatabaseMapping, String> mappingWithPattern : databaseMappingsForPattern.entrySet()) {
+          Callable<List<String>> callableTask = () -> {
+            try {
+              DatabaseMapping mapping = mappingWithPattern.getKey();
+              String pattern = mappingWithPattern.getValue();
+              List<String> databases = mapping.getClient().get_databases(pattern);
+              return getMappedWhitelistedDatabases(databases, mapping);
+            } catch (TException e) {
+              LOG.warn(errorMessage, e);
+              return null;
             }
-          } catch (TException e) {
-            LOG.warn("Can't fetch databases by pattern: {}", e.getMessage());
+          };
+          futures.add(executorService.submit(callableTask));
+        }
+
+        Iterator<DatabaseMapping> iterator = databaseMappingsForPattern.keySet().iterator();
+        for (Future<List<String>> future : futures) {
+          try {
+            DatabaseMapping mapping = iterator.next();
+            long timeout = mapping.getTimeout();
+            List<String> mappedDatabases = future.get(timeout, TimeUnit.MILLISECONDS);
+            combined.addAll(mappedDatabases);
+          } catch (InterruptedException | ExecutionException | TimeoutException | NullPointerException e) {
```

Comment:
We need to split these up:

- InteruptException

 probably ok to log.

- ExecutionException,

 I think this is the exception which we would normally get, so we should do what we did before which is something like:
```
          } catch (ExecutionException e) {
            LOG.warn("Can't set UGI: {}", e..getCause().getMessage());
          }
```

- TimeoutException:
We want to log something else to indicate that the metastore was to slow to respond and results are omitted.

How can we get a NullPointerException?

Reply:
From your comment above, 

> returning null is not great I think we'll get an error adding that to the list

That's how I got NullPointerException 😬
If we don't return null, we can stop worrying about it :)